pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -140,17 +142,55 @@ func readConfig(config io.Reader) (Config, error) {
 	return cfg, err
 }
 
+func readInstanceID() (string, error) {
+	// Try to find instance ID on the local filesystem (created by cloud-init)
+	const instanceIDFile = "/var/lib/cloud/data/instance-id"
+	idBytes, err := ioutil.ReadFile(instanceIDFile)
+	if err == nil {
+		instanceID := string(idBytes)
+		instanceID = strings.TrimSpace(instanceID)
+		glog.V(3).Infof("Got instance id from %s: %s", instanceIDFile, instanceID)
+		return instanceID, nil
+	}
+	glog.V(5).Infof("Cannot read %s: '%v', trying metadata server", instanceIDFile, err)
+
+	const instanceIDUrl = "http://169.254.169.254/2007-01-19/meta-data/instance-id"
+	resp, err := http.Get(instanceIDUrl)
+	if err != nil {
+		glog.V(3).Infof("Cannot read %s: %v", instanceIDUrl, err)
+		return "", err
+	}
+
```

Comment:
and maybe content-type? (I don't know the API well enough to know what is expected)